lib/full-site-editing/class-wp-rest-templates-controller.php

```diff
@@ -0,0 +1,545 @@
+		$changes->post_name = $request['slug'];
+		$result             = wp_insert_post( wp_slash( (array) $changes ), true );
+		if ( is_wp_error( $result ) ) {
+			return $result;
+		}
+		$id = $changes->theme . '|' . $changes->slug;
+
+		$template      = gutenberg_get_block_template( $id, $this->post_type );
+		$fields_update = $this->update_additional_fields_for_object( $template, $request );
+		if ( is_wp_error( $fields_update ) ) {
+			return $fields_update;
+		}
+
+		return $this->prepare_item_for_response(
+			gutenberg_get_block_template( $id, $this->post_type ),
+			$request
+		);
+	}
+
+	/**
+	 * Checks if a given request has access to delete a single template.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has dedlete access for the item, WP_Error object otherwise.
+	 */
+	public function delete_item_permissions_check( $request ) {
+		return $this->permissions_check( $request );
+	}
+
+	/**
+	 * Deletes a single template.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function delete_item( $request ) {
+		$template = gutenberg_get_block_template( $request['id'], $this->post_type );
+		if ( ! $template ) {
+			return new WP_Error( 'rest_template_not_found', __( 'No templates exists with that id.', 'gutenberg' ), array( 'status' => 404 ) );
+		}
+
+		$id    = $template->wp_id;
+		$force = (bool) $request['force'];
+
+		// If we're forcing, then delete permanently.
+		if ( $force ) {
+			$previous = $this->prepare_item_for_response( $template, $request );
+			wp_delete_post( $id, true );
+			$response = new WP_REST_Response();
+			$response->set_data(
+				array(
+					'deleted'  => true,
+					'previous' => $previous->get_data(),
+				)
+			);
+
+			return $response;
+		}
+
+		// Otherwise, only trash if we haven't already.
+		if ( 'trash' === $template->status ) {
+			return new WP_Error(
+				'rest_template_already_trashed',
+				__( 'The template has already been deleted.', 'gutenberg' ),
+				array( 'status' => 410 )
+			);
+		}
+
+		wp_trash_post( $id );
+		$template->status = 'trash';
+		return $this->prepare_item_for_response( $template, $request );
+	}
+
+	/**
+	 * Prepares a single template for create or update.
+	 *
+	 * @param WP_REST_Request $request Request object.
+	 * @return stdClass Changes to pass to wp_update_post.
+	 */
+	protected function prepare_item_for_database( $request ) {
+		$template           = $request['id'] ? gutenberg_get_block_template( $request['id'], $this->post_type ) : null;
+		$changes            = new stdClass();
+		$changes->post_name = $template->slug;
+		if ( null === $template ) {
+			$changes->post_type   = $this->post_type;
+			$changes->post_status = 'publish';
+			$changes->tax_input   = array(
+				'wp_theme' => wp_get_theme()->get_stylesheet(),
```

Comment:
Should we check the `$request` if a theme is set and use that (with a fallback to the current one)? To ensure this also works for themes that aren't currently active.